Source/utils/intrusive_optional.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+#include "appfat.h"
+#include "utils/stdcompat/optional.hpp"
+
+/// An optional that uses a field of the stored class and some value to store nullopt.
+#define DEFINE_INTRUSIVE_OPTIONAL_IMPL(OPTIONAL_CLASS, VALUE_CLASS, FIELD, NULL_VALUE, CONSTEXPR) \

```

Comment:
if this thing exploads one day is it generic enough that it's trivial to replace with the std optional?

Reply:
@AJenbo Yes, the API is a subset of `std::optional`, it even uses `std::nullopt` assignment etc.